### DISCUSSION TITLE
TypeScript Errors

### DISCUSSION THREAD

**Discussion by subhoghoshX** [photonstorm/phaser#6308] [Q&A] [answered]
## Version

* Phaser Version: 3.55.2+

* Operating system: any

* Browser: any

## Description

When trying to recreate the game using the website's [guide](https://phaser.io/tutorials/making-your-first-phaser-3-game/part8), I encountered a couple of TypeScript errors.

In a section while iterating the stars it shows `setBounceY` not found.

```js
const stars = this.physics.add.group({
  key: "star",
  repeat: 11,
  setXY: { x: 12, y: 0, stepX: 70 },
});

stars.children.iterate((child) => {
  child.setBounceY(Phaser.Math.FloatBetween(0.4, 0.8)); // Here it says `setBounceY` doesn't exist.
  return null;
});
```

![image](https://user-images.githubusercontent.com/108616679/205630123-9cc12fc8-85eb-4eda-a504-33dca7839c1e.png)

And just after that, in the tutorial, where it shows how to detect overlap, the doc shows to use `null` and `this` as the 3rd & 4th arguments.

![image](https://user-images.githubusercontent.com/108616679/205633306-cbe8cca8-888f-4e47-81cf-9d3abb776cf9.png)

But it shows a type error when I pass `null`.

![image](https://user-images.githubusercontent.com/108616679/205633397-b90bea78-cba3-4d81-9c73-7574439d9ad5.png)

Looking at the type it look like it expects something else:

![type screenshot](https://user-images.githubusercontent.com/108616679/205635741-8efa65d2-7792-44a4-8aea-1aeedd625ee7.png)

I encountered the above errors in phaser@3.55.2+ including the latest beta 16.

**Comment by subhoghoshX** [comment]
PS: I know creating an issue is easier than fixing it. I'm quite familiar with JS/TS, if anyone can point me in the right direction, I can try fixing it.

**Comment by photonstorm** [comment]
You need to cast 'child' in the iterate handler or TS won't understand what it is.

**Comment by subhoghoshX** [comment]
Cast 'child' to what type?

**Comment by photonstorm** [comment, marked as answer]
To what it is - an `Phaser.Physics.Arcade.Sprite`.

**Reply by subhoghoshX** [reply]
Ah that fixes it. Thanks much. There was no obvious way for me to know it.

**Reply by subhoghoshX** [reply]
One more question. What about the null error in the `this.physics.add.overlap()` function? It seems to me that the last two arguments are not needed.

**Reply by photonstorm** [reply]
You can set it to `undefined` instead of `null` to make the TS defs happy.

The final parameter is the scope of the callback, which is extremely important.

**Reply by subhoghoshX** [reply]
Yeah I also used `undefined` there. Didn't know the final parameter was so important! seemed to work without it. Thanks for clarifying.  Means a lot.

**Reply by photonstorm** [reply]
It sets the JS scope that the callback is invoked within. If you replace the handler with an arrow function, that will do the same thing (bound to the current scope, of course), so in that case, you can ignore it. But if you don't use an arrow function (because you wish to set the scope directly to a Game Object, for example), then that parameter is the easiest way to do it.

**Comment by Robouste** [comment]
Here is a proper solution to this issue:

```typescript

const stars = this.physics.add.group({
  key: "star",
  repeat: 11,
  setXY: { x: 12, y: 0, stepX: 70 },
});

stars.children.iterate((child) => {
  if (!(child instanceof Phaser.Physics.Arcade.Sprite)) {
    throw Error("Child is not an instance of Phaser.Physics.Arcade.Sprite");
  }

  child.setBounceY(Phaser.Math.FloatBetween(0.4, 0.8)); 

  return true;
});

```

No dirty casting, proper type handling. Phaser should really be more Typescript friendly.
